qiskit_nature/operators/second_quantization/spin_op.py

```diff
@@ -213,21 +214,34 @@ def __init__(
             data = [(data, 1)]
 
         if isinstance(data, list):
-            # [IXYZ]_index^power (power is optional) or [+-]_index
-            sparse = r"([IXYZ]_\d+(\^\d+)?|[\+\-]_\d+?)"
-            # space (\s) separated sparse label or dense label (repeat of [IXYZ+-])
-            label_pattern = re.compile(rf"^({sparse}\s)*{sparse}(?!\s)$|^[IXYZ\+\-]+$")
-            invalid_labels = [label for label, _ in data if not label_pattern.match(label)]
-            if invalid_labels:
-                raise ValueError(f"Invalid labels: {invalid_labels}")
+            if register_length is not None:  # Sparse label
+                # [IXYZ]_index^power (power is optional) or [+-]_index
+                sparse = r"([IXYZ]_\d+(\^\d+)?|[\+\-]_\d+?)"
+                # space (\s) separated sparse label
+                label_pattern = re.compile(rf"^({sparse}\s)*{sparse}(?!\s)$")
```

Comment:
Shall we compile this once outside of this class? We did something similar for the `VibrationOp`

Reply:
Why? I did so when I implemented `SpinOp` at first, but it does not change the performance, and I think it improves the readability. So I change it to current version.

Reply:
I just don't think it's necessary to recompile it every time if you can avoid it.
And having a variable called `SPARSE_LABEL_REGEX` (or something similar) does not hurt readability imo

Reply:
the cost of recompile is not so heavy and it requires. You need to go back when you read the code. It's easier to read if you can read it straightforward.

Reply:
I am fine with it either way :+1: